bsp/Src/BSP_SPI.c

```diff
@@ -0,0 +1,196 @@
+        stat = HAL_ERROR; // should this be error? should we add as many as we can?
+    }
+
+    return stat;
+}
+
+/**
+ * @brief Automatically called on transmit complete; recursively transmit
+ * SPI messages while we have messages in the tx queue to send.
+ * 
+ * @param hspi 
+ */
+void HAL_SPI_TxCpltCallback(SPI_HandleTypeDef *hspi) {
+    // This executes whenever a transmit completes
+    // We need this static second queue of lengths of messages we are sending so we can
+    // pop data in order appropriately.
+
+    // If the queue isn't empty...
+    uint8_t data_buf[LOCAL_BUFFER_SIZE];
+    if (!xQueueIsQueueEmptyFromISR(*_transmitQueuePtr) && !xQueueIsQueueEmptyFromISR(localTxMetaQueue)) {
+        uint16_t lengthToSend;
+        uint8_t temp;
+        xQueueReceiveFromISR(localTxMetaQueue, &lengthToSend, NULL); // get # of bytes to keep transmitting
+        for (uint16_t i = 0; i < lengthToSend && i < LOCAL_BUFFER_SIZE; i++) {
+            xQueueReceiveFromISR(*_transmitQueuePtr, &(data_buf[i]), NULL);
+        }
+        HAL_SPI_Transmit_IT(hspi, data_buf, lengthToSend); // recurse
+    }
+}
+
+/**
+ * @brief Perform a SPI read
+ * 
+ * @param spiHandle The SPI handle to read from
+ * @return HAL_StatusTypeDef 
+ */
+static uint8_t spiRxTempBuffer[LOCAL_BUFFER_SIZE] = {0};
+HAL_StatusTypeDef BSP_SPI_Read(SPI_HandleTypeDef* spiHandle, uint16_t len) {
+    HAL_StatusTypeDef stat;
+    
+    if (len <= LOCAL_BUFFER_SIZE) {
+        stat = HAL_SPI_Receive_IT(spiHandle, spiRxTempBuffer, len);
+        xQueueSend(localRxMetaQueue, &len, 0); // push metadata (length)
+    } else {
+        stat = HAL_ERROR;
+    }
+
+    return stat;
+}
+
+/**
+ * @brief Automatically executed when a SPI read is finished; copy
+ * all data from the read buffer in SPI_Read() into the rxQueue passed
+ * in by the init function
+ * 
+ * @param hspi 
+ */
+// This specifically needs review. I think I might have some weird logic between
+// Read() and this callback.
+void HAL_SPI_RxCpltCallback(SPI_HandleTypeDef *hspi) {
+    uint16_t toSend;
+    xQueueReceiveFromISR(localRxMetaQueue, &toSend, NULL);
+    for (uint16_t i = 0; i < toSend; i++) {
+        xQueueSendFromISR(*_receiveQueuePtr, &(spiRxTempBuffer[i]), 0);
+    }
+}
+
+/**
+ * @brief Sets SPI chip select.
+ * 
+ * This function needs to be updated once the leaderSOM board is finalized and SPI CS is set.
+ * 
+ * @param spiHandle SPI handle to modify
+ * @return HAL_StatusTypeDef 
+ */
+void BSP_SPI_SetStateCS(SPI_HandleTypeDef* spiHandle, uint8_t val) {
+    uint8_t hal_val = (val) ? GPIO_PIN_SET : GPIO_PIN_RESET;
+    // BSP config only supports SPI2/SPI3
+    if (spiHandle->Instance == SPI2) {
+        // SPI2 NSS is on PA11, AF5
+        HAL_GPIO_WritePin(GPIOA, GPIO_PIN_11, hal_val);
+    } else if (spiHandle->Instance == SPI3) {
+        // SPI3 NSS is on PA4, AF6
+        HAL_GPIO_WritePin(GPIOA, GPIO_PIN_4, hal_val);
+    }
+}
```

Comment:
I want to be able to change the chip select pin and not have it hardcoded. Technically chip select can be GPIO whatever pin.